r/geoarrow/src/r-vctr.c

```diff
@@ -0,0 +1,116 @@
+
+#define R_NO_REMAP
+#include <R.h>
+#include <Rinternals.h>
+
+#include "geoarrow.h"
+
+SEXP geoarrow_c_vctr_chunk_offsets(SEXP array_list) {
+  int num_chunks = Rf_length(array_list);
+  SEXP offsets_sexp = PROTECT(Rf_allocVector(INTSXP, num_chunks + 1));
+  int* offsets = INTEGER(offsets_sexp);
+  offsets[0] = 0;
+  int64_t cumulative_offset = 0;
+
+  struct ArrowArray* array;
+  for (int i = 0; i < num_chunks; i++) {
+    array = (struct ArrowArray*)R_ExternalPtrAddr(VECTOR_ELT(array_list, i));
+    cumulative_offset += array->length;
+    if (cumulative_offset > INT_MAX) {
+      Rf_error("Can't build geoarrow_vctr with length > INT_MAX");
+    }
+
+    offsets[i + 1] = cumulative_offset;
+  }
+
+  UNPROTECT(1);
+  return offsets_sexp;
+}
+
+static int resolve_chunk(int* sorted_offsets, int index, int start_offset_i,
+                         int end_offset_i) {
+  if (start_offset_i >= (end_offset_i - 1)) {
+    return start_offset_i;
+  }
+
+  int mid_offset_i = start_offset_i + (end_offset_i - start_offset_i) / 2;
+  int mid_index = sorted_offsets[mid_offset_i];
+  if (index < mid_index) {
+    return resolve_chunk(sorted_offsets, index, start_offset_i, mid_offset_i);
+  } else {
+    return resolve_chunk(sorted_offsets, index, mid_offset_i, end_offset_i);
+  }
+}
+
+SEXP geoarrow_c_vctr_chunk_resolve(SEXP indices_sexp, SEXP offsets_sexp) {
+  int* offsets = INTEGER(offsets_sexp);
+  int end_offset_i = Rf_length(offsets_sexp) - 1;
+  int last_offset = offsets[end_offset_i];
+
+  int n = Rf_length(indices_sexp);
+  SEXP chunk_indices_sexp = PROTECT(Rf_allocVector(INTSXP, n));
+  int* chunk_indices = INTEGER(chunk_indices_sexp);
+
+  int buf[1024];
+  for (int i = 0; i < n; i++) {
+    if (i % 1024 == 0) {
+      INTEGER_GET_REGION(indices_sexp, i, 1024, buf);
+    }
+    int index0 = buf[i % 1024];
+
+    if (index0 < 0 || index0 > last_offset) {
+      chunk_indices[i] = NA_INTEGER;
+    } else {
+      chunk_indices[i] = resolve_chunk(offsets, index0, 0, end_offset_i);
+    }
+  }
+
+  UNPROTECT(1);
+  return chunk_indices_sexp;
+}
+
+SEXP geoarrow_c_vctr_as_slice(SEXP indices_sexp) {
+  if (TYPEOF(indices_sexp) == INTSXP) {
```

Comment:
Inverting this condition & early return reduces the nesting here
```suggestion
  if (TYPEOF(indices_sexp) != INTSXP) return R_NilValue;
```